vowpalwabbit/object_pool.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include <set>
+#include <queue>
+
+namespace VW
+{
+
+// Not thread safe
+template <typename T, typename TInitializer>
+struct object_pool
+{
+  object_pool() = default;
+  object_pool(size_t initial_chunk_size, TInitializer initializer = {}, size_t chunk_size = 8)
+    : m_initializer(initializer),
+    m_initial_chunk_size(initial_chunk_size),
+    m_chunk_size(chunk_size)
+  {
+    new_chunk(initial_chunk_size);
+  }
+
+  void return_object(T* obj)
+  {
+    assert(is_from_pool(obj));
+    m_pool.push(obj);
+  }
+
+  T* get_object()
+  {
+    if (m_pool.empty())
+    {
+      new_chunk(m_chunk_size);
+    }
+
+    auto obj = m_pool.front();
+    m_pool.pop();
+    return obj;
+  }
+
+  bool empty() const { return m_pool.empty(); }
+
+  size_t available() const { return m_pool.size(); }
+
+  size_t size() const {
+    size_t size = 0;
+    auto num_chunks = m_chunk_bounds.size();
+
+    if(m_chunk_bounds.size() > 0 && m_initial_chunk_size > 0)
+    {
+      size += m_initial_chunk_size;
+      num_chunks--;
+    }
+
+    size += num_chunks * m_chunk_size;
+    return size;
+  }
+
+  bool is_from_pool(T* obj) const {
+    for(auto& bound : m_chunk_bounds)
+    {
+      if(obj >= bound.first && obj <= bound.second)
+      {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+ private:
+  void new_chunk(size_t size)
+  {
+    if(size == 0)
+    {
+      return;
+    }
+
+    m_chunks.push_back(std::unique_ptr<T[]>(new T[size]));
```

Comment:
Why is unique_ptr necessary here?

Reply:
So that memory is automatically freed for us on destruction. If we can avoid writing delete then it's a win :)

Reply:
I understand what unique_ptr does.  The cost (I believe) is that you end up double boxing things making every pointer deref extra painful.  Avoiding unnecessary boxing is one of the first things I look for in making code run faster.

Reply:
These unique_ptrs are never accessed except for on creation of a chunk and at deletion time, so extra boxing shouldn't affect things here. The raw pointers are returned for normal use

Reply:
So you can remove dependence on unique_ptr and call delete explicitly without issues?  That seems like a plausible win in terms of reducing the amount of code.

Reply:
`delete[]` would need to be called on the memory block pointer which `unique_ptr` is managing. This class passes out pointers to individual pieces within this, so the delete cannot be done on these individual pointers.